gm4_dripleaf_filters/data/gm4_dripleaf_filters/functions/check_hopper_filter.mcfunction

```diff
@@ -0,0 +1,35 @@
+# @s item on a dripleaf with a hopper to the side of the dripleaf containing at least one item
+# run from on_leaf
+
+# store the item's ID in storage
+data modify storage gm4_dripleaf_filters:cache itemCompare set from entity @s Item.id
+# store the hopper's items in storage
+data modify storage gm4_dripleaf_filters:cache hopperCompare set from block ^ ^-1 ^1 Items
+
+# loop through the slots to see if any IDs match
+# slot 0
+execute store success storage gm4_dripleaf_filters:cache itemsAreDifferent byte 1 run data modify storage gm4_dripleaf_filters:cache hopperCompare[0].id set from storage gm4_dripleaf_filters:cache itemCompare
+execute if data storage gm4_dripleaf_filters:cache {itemsAreDifferent:0b} run tag @s add gm4_dripleaf_filters_match

```

Comment:
Many of these checks are using tags on the entity, but this can be moved completely to scores of fake players, which is more efficient. For an example, look at how the checks are done in custom crafters recipe_validity_check (lib_machines version). This function checks if the counts match, but the same can be done for the id. https://github.com/BluePsychoRanger/GM4_Datapacks/blob/lib-machines/lib_machines/data/gm4_custom_crafters-2.0/functions/recipe_validity_check.mcfunction

Reply:
I have now moved the two tags and the byte storage value to custom players with at $ prefix. Thanks for the feedback!